src/languages/haskell.js

```diff
@@ -62,6 +62,42 @@ export default function(hljs) {
     contains: LIST.contains
   };
 
+  /* See:
+
+     - https://www.haskell.org/onlinereport/lexemes.html
+     - https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/binary_literals.html
+     - https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/numeric_underscores.html
+     - https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/hex_float_literals.html
+
+  */
+  const decimalDigits = '([0-9]_*)+';
+  const hexDigits = '([0-9a-gA-F]_*)+';
+  const binaryDigits = '([01]_*)+';
+  const octalDigits = '([0-7]_*)+';
+
+  const NUMBER = {
+    className: 'number',
+    relevance: 0,
+    variants: [
+      // decimal floating-point-literal (subsumes decimal-literal)
+      {
+        match: `(-?)\\b(${decimalDigits})(\\.(${decimalDigits}))?` + `([eE][+-]?(${decimalDigits}))?\\b`
+      },
+      // hexadecimal floating-point-literal (subsumes hexadecimal-literal)
+      {
+        match: `(-?)\\b0x_*(${hexDigits})(\\.(${hexDigits}))?` + `([pP][+-]?(${decimalDigits}))?\\b`
+      },
+      // octal-literal
+      {
+        match: `(-?)\\b0b(${octalDigits})\\b`
```

Comment:
Is this correct? From manual: 

> octal (prefixed by 0o or 0O)

Reply:
Do we have no octal tests that would catch this mistake?

Reply:
You're right, that's a copy+paste error. 

I didn't know about the `[oO]`, but that seems to be a rule across all notations (`[xX]`, `[bB]`). I've updated the patterns accordingly.